src/scheduling/cost_models/simulated_quincy_cost_model.cc

```diff
@@ -0,0 +1,358 @@
+      // case where block is stored in two machines in same rack
+      uint64_t frequency = FindWithDefault(rack_frequency, rack, 0);
+      rack_frequency[rack] = frequency + num_blocks;
+    }
+  }
+
+  preferred_machine_map_[task_id] = new ResourceCostMap_t();
+  auto &preferred_machines = *(preferred_machine_map_[task_id]);
+  for (auto freq : machine_frequency) {
+  	SimulatedDFS::NumBlocks_t num_local_blocks = freq.second;
+  	double proportion = num_local_blocks;
+  	proportion /= total_num_blocks;
+  	if (proportion >= proportion_machine_preferred_) {
+  		// add to preferred list and compute cost
+  		ResourceID_t machine = freq.first;
+  		EquivClass_t rack = machine_to_rack_map_[machine];
+  		SimulatedDFS::NumBlocks_t num_rack_blocks = rack_frequency[rack];
+
+  		// local blocks are charged at 0,
+  		// blocks transferred between nodes in rack charged at tor_transfer_cost_,
+  		// blocks between racks charged at rack_transfer_cost_
+  		// Totals so far are inclusive, calculate exclusive ones
+  		num_rack_blocks -= num_local_blocks;
+  		SimulatedDFS::NumBlocks_t num_core_blocks =
+			                    total_num_blocks - num_rack_blocks - num_local_blocks;
+
+  		Cost_t cost = num_core_blocks * core_transfer_cost_;
+  		cost += num_rack_blocks * tor_transfer_cost_;
+
+  		preferred_machines[machine] = cost;
+
+  		VLOG(2) << "Task " << task_id << " preferred machine " << machine
+  		        << " cost " << cost << "; proportion " << proportion
+  		        << ", local " << num_local_blocks << " rack " << num_rack_blocks
+  		        << " total " << total_num_blocks;
+  	}
+  }
+
+  preferred_rack_map_[task_id] = unordered_map<EquivClass_t, Cost_t>();
+  auto &preferred_racks = preferred_rack_map_[task_id];
+  for (auto freq : rack_frequency) {
+  	SimulatedDFS::NumBlocks_t num_rack_blocks = freq.second;
+  	double proportion = num_rack_blocks;
+  	proportion /= total_num_blocks;
+  	if (proportion > proportion_rack_preferred_) {
+  		EquivClass_t rack = freq.first;
+
+  		SimulatedDFS::NumBlocks_t num_core_blocks = total_num_blocks
+  				 	 	 	 	 	 	 	 	 	 	 	 	 	 	 	 	 	  - num_rack_blocks;
+
+  		Cost_t cost = num_core_blocks * core_transfer_cost_;
+  		cost += num_rack_blocks * tor_transfer_cost_;
+
+  		preferred_racks[rack] = cost;
+
+  		VLOG(2) << "Task " << task_id << " preferred rack " << rack;
+  	}
+  }
+
+  Cost_t cost = total_num_blocks * core_transfer_cost_;
+  cluster_aggregator_cost_[task_id] = cost;
+}
+
+void SimulatedQuincyCostModel::AddTask(TaskID_t task_id) {
+	BuildTaskFileSet(task_id);
+	ComputeCostsAndPreferredSet(task_id);
+}
+
+void SimulatedQuincyCostModel::RemoveTask(TaskID_t task_id) {
+  delete preferred_machine_map_[task_id];
+
+  file_map_.erase(task_id);
+  preferred_machine_map_.erase(task_id);
+  preferred_rack_map_.erase(task_id);
+	cluster_aggregator_cost_.erase(task_id);
+}
+
+FlowGraphNode* SimulatedQuincyCostModel::GatherStats(FlowGraphNode* accumulator,
```

Comment:
I am a bit surprised that these methods are not implemented. I think we should use them in order to adjust the costs of the preference arcs. We should gather the number of tasks that are running on a machine. We don't want to place N tasks on the same machine if they are all hitting hard the disk.